### PR TITLE
[JENKINS-72978] Remove executors widget from Dashboard and nodes page

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-72978](https://issues.jenkins.io/browse/JENKINS-72978).

Implements https://matrix.to/#/!HKutvjxPnajVyCNLhF:matrix.org/$kfLAa91tXfp6Gv10PgoY8i7GM3GgqfDG_pz39-nWsDQ?via=gitter.im&via=matrix.org&via=mozilla.org

Adds a replacement page with the same information. I thought about redesigning it but I didn't think it made sense to do in this pull request.

image

This could fairly easily be detached to a plugin if people really want this still.

### Testing done

manually checked the widget doesn't exist anymore except on the computer page

### Proposed changelog entries

- JENKINS-72978, Removed executor status from dashboard, executor status can now be viewed from Manage Jenkins -> Nodes -> Status

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Comment by NotMyFault** [comment]
This is indeed pretty neat. Would it be useful, for a limited period of time, to display a link in the position of the current executor widget, that takes you to the new overview?

**Comment by timja** [comment]
> This is indeed pretty neat. Would it be useful, for a limited period of time, to display a link in the position of the current executor widget, that takes you to the new overview?

I think changelog is probably ok, I've added a link to the nodes page for low privileged users as there was no way for them to get there now

**Comment by mawinter69** [comment]
A major drawback is that the executors widget is now also removed from individual agents.

**Comment by mawinter69** [comment]
Maybe we can just remove the Factories for views `ViewFactoryImpl` and ComputerSet `ComputerSetFactoryImpl`, Can we then reuse the t:executors on the new status page? The main task of showing the running builds is done by `executorCell.jelly` files which can be found in core and durable-task-step plugin, so in order to redesign it might require changes in that plugin as well. (See also #8459)

**Comment by timja** [comment]
> Maybe we can just remove the Factories for views `ViewFactoryImpl` and ComputerSet `ComputerSetFactoryImpl`

Makes sense

> Can we then reuse the t:executors on the new status page? 

What's the reason to reuse it on the status page? Its current setup assumes its in the sidepanel and likely needs some changes to make it fit in properly, I've done a minimal bit so that it looks okay for now though.

**Comment by mawinter69** [comment]
If we keep the widget for computers and also have the status page we would duplicate the code. And the differences are minimal.
Using 
```
    <l:main-panel>
      <l:app-bar title="${%Status}"/>
      <div id="side-panel">
      <t:executors/>
      </div>
    </l:main-panel>
```
also works now, just a bit different styling.
image

Another aspect is the auto refresh, which is missing (also doesn't work with my approach but only because the url to fetch the data is wrong)

**Comment by timja** [comment]
> Another aspect is the auto refresh, which is missing (also doesn't work with my approach but only because the url to fetch the data is wrong)

Yeah I wasn't sure if we wanted that or not, can add it though if worth it. There's a smaller risk of people leaving it open forever and lots of polling updates. 

> also works now, just a bit different styling.

Yeah styling is a bit messed up though I had to make some minor changes to the markup to make it work like removing `float:right` and a couple of other things

File: core/src/main/resources/hudson/model/View/sidepanel.jelly
Comment (mawinter69) [inline core/src/main/resources/hudson/model/View/sidepanel.jelly:61]: the permission check here `it.READ` means that you need either `Item/Read` or `View/Read` permission to be able to see the link. In general not bad though not really protecting the `/computer` url which only requires `Overall/Read`

From an admin point of view it means now 3 clicks to get from the root to the executor status (Manage Jenkins -> Nodes -> Status) , before it was just one click.
And an administrator will much more likely need to go there than any other user (I've been using the direct link from the widget quite often lately). So Maybe have that as a link to everyone and not just non admins.
As a workaround one could use the sidebarlink plugin (or with one more click but then on every page with customizable header plugin)

Another possibility might be to have the widget only for admins (one can also have the factory that adds the widget to views in a plugin and with a config option to limit it to admins)

**Review by mawinter69 — COMMENTED** [review]
One comment, not really a blocker but something to think about

File: core/src/main/resources/hudson/model/View/sidepanel.jelly
Comment (timja) [reply core/src/main/resources/hudson/model/View/sidepanel.jelly:61]:
```suggestion
        <l:task href="${rootURL}/computer" icon="symbol-computer" title="${%Nodes}" />
```

Comment (timja) [reply core/src/main/resources/hudson/model/View/sidepanel.jelly]: I also use the direct link but I'm a bit wary of adding it to the View sidepanel as we're trying to remove links from there, and not just clutter it up.

You can get to it from the context menu without much effort:

image

**Review by mawinter69 — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by daniel-beck** [comment]
I'd like an opportunity to review :) Thanks!

**Comment by daniel-beck** [comment]
What's the intended level of done-ness of this PR? This looks like a reasonable first step and implements the high-level request, but doesn't go much further than that. For example https://github.com/jenkinsci/jenkins/blob/c660d271afd8fdd52d67cd76fd26a1a5a1f8a94b/core/src/main/java/hudson/model/View.java#L158-L161 is now obsolete and probably should be made `@Deprecated`, `transient` and `Boolean`.

Some less clear interaction issues:

1. It's difficult to get from a blocked build in the queue to the UI that allows fixing the blockage, now that the widgets aren't stacked right on top of each other. The hourglass icon is the only added UI element, and it just links to https://www.jenkins.io/redirect/troubleshooting/executor-starvation
2. It's unclear what the expected flow is for an admin to (wait to) restart Jenkins once it's safe to do so.

> This could fairly easily be detached to a plugin if people really want this still.

IMO most of the information being presented should remain accessible to users who administer Jenkins. I don't think anyone cares about which # executor a build is running on, but the rest seems interesting. The widget on the computer view answers the question, which job is currently using this node; while the new Status page does the same across all nodes (like the widget did before) and e.g. tells admins how far a potential "quiet down" has been successful.

Thoughts about delivering this as a UI experiment that users can opt out of, just in case? (Then of course without removal of `filterExecutors` for now.)

**Comment by mawinter69** [comment]
Independent of this change I think the widget could be improved by not showing idle executors. I have a few agents that perform small jobs that usually take not more than 1 minute but sometimes they can also run much longer. So these agents have 15 executors but most of the time only 1 or 2 are busy if at all. The idle executors don't need to be shown here, instead we could add the # of busy/total executors after the agent name
Something like this:
image

**Comment by mawinter69** [comment]
What would be also helpful is if one could collapse the executors of a single agent like you can collapse the complete widget

**Comment by daniel-beck** [comment]
> not showing idle executors

Basically  what I mentioned as

> I don't think anyone cares about which # executor a build is running on

The specific executor assignment wasn't even useful information when it was in the sidepanel, only used/total would be relevant IMO. But that's an easy followup (or even independent parallel) change to make.

**Comment by mawinter69** [comment]
created #9177 with the idle executors removed

**Review by janfaracik — APPROVED** [review]

**Comment by janfaracik** [comment]
Minor nitpick/not blocking - the widget is inset a little bit`margin-left: 10px` - can be resolved as part of a cleanup of that view in the future.

**Comment by ArieHein** [comment]
Please do not remove the status pane by default. 

Admins want to see immediate status of nodes, and not have to do more actions to get to overall status.
Users want to see their own pipelines running status and not care about 200 other executors.

Something that is core to the tool should not be a plugin and i dare say a widget.

having ALSO a status page,, especially is a multi-executor environment is ok, but not instead of front page side panel
For admins in such an environment you might show current active (to limit the amount of scrolling, place a max no matter what) or use favorites where admins can point to the most critical one or ones they care more about.

**Comment by mawinter69** [comment]
> Please do not remove the status pane by default.
> 
> Admins want to see immediate status of nodes, and not have to do more actions to get to overall status. Users want to see their own pipelines running status and not care about 200 other executors.
> 
> Something that is core to the tool should not be a plugin and i dare say a widget.
> 
> having ALSO a status page,, especially is a multi-executor environment is ok, but not instead of front page side panel For admins in such an environment you might show current active (to limit the amount of scrolling, place a max no matter what) or use favorites where admins can point to the most critical one or ones they care more about.

It should be easy to create a plugin that will restore the widget, ideally as a per user setting

**Pull request closed without merging**
